lib/aot/compiler.cpp

```diff
@@ -2905,6 +3171,148 @@ class FunctionCompiler {
     stackPush(PHIRet);
   }
 
+  void compileAtomicCheckOffsetAlignment(llvm::Value *Offset,
+                                         llvm::IntegerType *IntType) {
+    const auto BitWidth = IntType->getBitWidth();
+    auto *BWMask = llvm::ConstantInt::get(Context.Int64Ty, (BitWidth >> 3) - 1);
+    auto *Value = Builder.CreateAnd(Offset, BWMask);
+    auto *OkBB = llvm::BasicBlock::Create(LLContext, "address_align_ok", F);
+    auto *IsAddressAligned =
+        createLikely(Builder, Builder.CreateICmpEQ(Value, Builder.getInt64(0)));
+    Builder.CreateCondBr(IsAddressAligned, OkBB,
+                         getTrapBB(ErrCode::UnalignedAtomicAccess));
+
+    Builder.SetInsertPoint(OkBB);
+  }
+
+  void compileMemoryFence() {
+    Builder.CreateFence(llvm::AtomicOrdering::SequentiallyConsistent);
+  }
+  void compileAtomicNotify() {
+    Builder.CreateCall(Context.getIntrinsic(
+                           Builder,
+                           AST::Module::Intrinsics::kMemoryAtomicNotify,
+                           llvm::FunctionType::get(Context.VoidTy, {}, false)),
+                       {});
+  }
+  void compileAtomicWait(unsigned MemIdx, unsigned BitWidth) {
+    auto *Offset = Builder.CreateSExtOrTrunc(stackPop(), Context.Int32Ty);
+    auto *ExpectedValue =
+        Builder.CreateSExtOrTrunc(stackPop(), Context.Int64Ty);
+    auto *Timeout = Builder.CreateSExtOrTrunc(stackPop(), Context.Int64Ty);
+
+    stackPush(Builder.CreateCall(
+        Context.getIntrinsic(
+            Builder, AST::Module::Intrinsics::kMemoryAtomicWait,
+            llvm::FunctionType::get(Context.Int32Ty,
+                                    {Context.Int32Ty, Context.Int32Ty,
+                                     Context.Int64Ty, Context.Int64Ty,
+                                     Context.Int32Ty},
+                                    false)),
+        {Builder.getInt32(MemIdx), Offset, ExpectedValue, Timeout,
+         Builder.getInt32(BitWidth)}));
+  }
+  void compileAtomicLoad(unsigned MemoryIndex, unsigned Alignment,
+                         llvm::IntegerType *IntType,
+                         llvm::IntegerType *TargetType, bool Signed = false) {
+
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
```

Comment:
Final offset should add offset value in `instruction`.
offset = 'stack pop value' + 'instruction offset value'